src/Sarif.Converters/ContrastSecurityConverter.cs

```diff
@@ -0,0 +1,1364 @@
+        {
+            if (input == null)
+            {
+                throw (new ArgumentNullException(nameof(input)));
+            }
+
+            if (output == null)
+            {
+                throw (new ArgumentNullException(nameof(output)));
+            }
+
+            LogicalLocations.Clear();
+
+            _files = new HashSet<FileData>(FileData.ValueComparer);
+
+            var context = new ContrastLogReader.Context();
+
+            // 1. First we initialize our global rules data from the SARIF we have embedded as a resource
+            Assembly assembly = typeof(ContrastSecurityConverter).Assembly;
+            SarifLog sarifLog;
+
+            using (var stream = assembly.GetManifestResourceStream(ContrastSecurityConverter.ContrastSecurityRulesData))
+            using (var streamReader = new StreamReader(stream))
+            {
+                string prereleaseRuleDataLogText = streamReader.ReadToEnd();
+                PrereleaseCompatibilityTransformer.UpdateToCurrentVersion(prereleaseRuleDataLogText, forceUpdate: true, Newtonsoft.Json.Formatting.Indented, out string currentRuleDataLogText);
+                sarifLog = JsonConvert.DeserializeObject<SarifLog>(currentRuleDataLogText);
+            }
+
+            // 2. Retain a pointer to the rules dictionary, which we will use to set rule severity
+            Run run = sarifLog.Runs[0];
+            _rules = run.Resources.Rules.ToDictionary(rule => rule.Id);
+
+            run.OriginalUriBaseIds = new Dictionary<string, FileLocation>
+            {
+                {  "SITE_ROOT", new FileLocation { Uri = new Uri(@"E:\src\WebGoat.NET") } } 
+            };
+
+            // 3. Now, parse all the contrast XML to create the complete results set
+            var results = new List<Result>();
+            var reader = new ContrastLogReader();
+            reader.FindingRead += (ContrastLogReader.Context current) => { results.Add(CreateResult(current)); };
+            reader.Read(context, input);
+
+            // 4. Construct the files array, based on all results returned
+            var fileInfoFactory = new FileInfoFactory(MimeType.DetermineFromFileExtension, dataToInsert);
+            HashSet<FileData> files = fileInfoFactory.Create(results);
+
+            // 5. Finally, complete the SARIF log file with various tables and then the results
+            output.Initialize(run);
+
+            foreach (FileData fileData in _files)
+            {
+                files.Add(fileData);
+            }
+
+            if (files?.Any() == true)
+            {
+                output.WriteFiles(files.ToList());
+            }
+
+            if (LogicalLocations?.Any() == true)
+            {
+                output.WriteLogicalLocations(LogicalLocations);
+            }
+
+            //if (_rules?.Any() == true)
+            //{
+            //    output.WriteRules(_rules.Values.ToList());
+            //}
```

Comment:
>//} [](start = 12, length = 3)

With this code uncommented, the rules were being persisted twice. I need to find out why.